ksql-common/src/main/java/io/confluent/ksql/function/GenericsUtil.java

```diff
@@ -0,0 +1,256 @@
+ */
+
+package io.confluent.ksql.function;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.KsqlPreconditions;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.apache.kafka.connect.data.Field;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+
+public final class GenericsUtil {
+
+  private static final String PREFIX = "<";
+  private static final String SUFFIX = ">";
+  private static final Pattern GENERIC_PATTERN = Pattern.compile("<(?<name>.*)>");
+
+  private GenericsUtil() { }
+
+  /**
+   * @param typeName  the generic type name (e.g. {@code T})
+   * @return a {@link SchemaBuilder} for a generic type with that name
+   */
+  public static SchemaBuilder generic(final String typeName) {
+    return SchemaBuilder.bytes().optional().name(PREFIX + typeName + SUFFIX);
+  }
+
+  /**
+   * @param typeName  the generic type name (e.g. {@code T})
+   * @return a {@link SchemaBuilder} for a generic array with generic element type
+   */
+  public static SchemaBuilder array(final String typeName) {
+    return SchemaBuilder.array(generic(typeName).build()).optional();
+  }
+
+  /**
+   * @param keyType       the type for the map key
+   * @param valueTypeName the generic type name (e.g. {@code T}) for the map value
+   * @return a {@link SchemaBuilder} for a map with {@code keyType} keys and generic value type
+   */
+  public static SchemaBuilder map(final Schema keyType, final String valueTypeName) {
+    return SchemaBuilder.map(keyType, generic(valueTypeName).build()).optional();
+  }
+
+  /**
+   * @param type  the type
+   * @return whether or not {@code type} is a generic type
+   * @apiNote container generics (e.g. {@code ARRAY<T>}) will return {@code false},
+   *          use {@link #constituentGenerics(Schema)}
+   */
+  public static boolean isGeneric(final Schema type) {
+    return type.name() != null
+        && type.name().startsWith(PREFIX)
+        && type.name().endsWith(SUFFIX);
+  }
+
+  /**
+   * @param type  the type
+   * @return all generics contained within the type, for example: {@code Map<K, V>} would return
+   *         a set containing {@code {<K>, <V>}}
+   */
+  public static Set<Schema> constituentGenerics(final Schema type) {
+    switch (type.type()) {
+      case ARRAY:
+        return constituentGenerics(type.valueSchema());
+      case MAP:
+        return Sets.union(
+            constituentGenerics(type.keySchema()),
+            constituentGenerics(type.valueSchema()));
+      case STRUCT:
```

Comment:
This is implemented because it's called to determine whether or not something has generics, which is used to determine whether or not to call `identifyGenerics`. Throwing an exception here would break the normal flow (this method is called for every KsqlFunction)

Reply:
ah yeah that makes perfect sense. thanks!